contracts/viewers/FLIRebalanceViewer.sol

```diff
@@ -0,0 +1,253 @@
+ * This contract can only determine whether to use Uniswap V3 or Uniswap V2 (or forks) for rebalancing. Since AMMTradeSplitter adheres to
+ * the Uniswap V2 router interface, this contract is compatible with that as well.
+ */
+contract FLIRebalanceViewer {
+
+    using StringArrayUtils for string[];
+    using SafeMath for uint256;
+    using PreciseUnitMath  for uint256;
+
+    /* ============ State Variables ============ */
+
+    IFLIStrategyExtension public fliStrategyExtension;
+
+    IQuoter public uniswapV3Quoter;
+    IUniswapV2Router public uniswapV2Router;
+
+    string public uniswapV3ExchangeName;
+    string public uniswapV2ExchangeName;
+
+    /* ============ Constructor ============ */
+
+    /**
+     * Sets state variables
+     *
+     * @param _fliStrategyExtension     FlexibleLeverageStrategyAdapter contract address
+     * @param _uniswapV3Quoter          Uniswap V3 Quoter contract address
+     * @param _uniswapV2Router          Uniswap v2 Router contract address
+     * @param _uniswapV3ExchangeName    Name of Uniswap V3 exchange in Set's IntegrationRegistry (ex: UniswapV3ExchangeAdapter)
+     * @param _uniswapV2ExchangeName    Name of Uniswap V2 exchange in Set's IntegrationRegistry (ex: AMMSplitterExchangeAdapter)
+     */
+    constructor(
+        IFLIStrategyExtension _fliStrategyExtension,
+        IQuoter _uniswapV3Quoter,
+        IUniswapV2Router _uniswapV2Router,
+        string memory _uniswapV3ExchangeName,
+        string memory _uniswapV2ExchangeName
+    )
+        public
+    {
+        fliStrategyExtension = _fliStrategyExtension;
+        uniswapV3Quoter = _uniswapV3Quoter;
+        uniswapV2Router = _uniswapV2Router;
+        uniswapV3ExchangeName = _uniswapV3ExchangeName;
+        uniswapV2ExchangeName = _uniswapV2ExchangeName;
+    }
+
+    /* =========== External Functions ============ */
+
+    /**
+     * Gets the priority order for which exchange should be used while rebalancing. Mimics the interface for
+     * shouldRebalanceWithBound of FlexibleLeverageStrategyExtension. Note: this function is not marked as view
+     * due to a quirk in the Uniswap V3 Quoter contract, but should be static called to save gas
+     *
+     * @param _customMinLeverageRatio       Min leverage ratio passed in by caller
+     * @param _customMaxLeverageRatio       Max leverage ratio passed in by caller
+     *
+     * @return string[] memory              Ordered array of exchange names to use. Earlier elements in the array produce the best trades
+     * @return ShouldRebalance[] memory     Array of ShouldRebalance Enums. Ordered relative to returned exchange names array
+     */
+    function shouldRebalanceWithBounds(
+        uint256 _customMinLeverageRatio,
+        uint256 _customMaxLeverageRatio
+    )
+        external
+        returns(string[] memory, FlexibleLeverageStrategyExtension.ShouldRebalance[] memory)
+    {
+
+        string[] memory enabledExchanges = fliStrategyExtension.getEnabledExchanges();
```

Comment:
Feel like we can dedupe some code here? We dont need to fetch enabled exchanges since line 225 (shouldRebalanceWithBounds) also returns the list of exchanges. We could set enabledExchanges to that first array and bring it up to this level

Reply:
Potentially one other way to organize the logic is to call FlexibleLeverageStrategyAdapter.shouldRebalanceWithBounds() first, then take the first return array (exchangeNames) and pass that into getChunkRebalanceNotional(), and then we get the `indexOf` the exchangeNames. This way we have all the data we need upfront to fetch quote and return into a priority order